### ISSUE TITLE
Operation is not valid due to the current state of the object

### ISSUE THREAD

**Issue by darko1979** [linq2db/linq2db#3226]
I have following issue. Following two queries work fine:

```c#
            var items1 = await db.Items
                .OrderBy(x => x.Values.Sum(y => y.Value))
                .Select(x => new {
                    Id = x.Id,
                    Text = x.Text
                })
                .ToListAsync();
            Console.WriteLine($"items1: {JsonSerializer.Serialize(items1)}");

            var items2 = await db.Items
                .Select(x => new {
                    Id = x.Id,
                    Text = x.Text,
                    Summary = x.Values.Select(y => new { Total = y.Value }).AsEnumerable()
                })
                .ToListAsync();
            Console.WriteLine($"items2: {JsonSerializer.Serialize(items2)}");
```
But following:
```c#
            var items3 = await db.Items
                .OrderBy(x => x.Values.Sum(y => y.Value))
                .Select(x => new {
                    Id = x.Id,
                    Text = x.Text,
                    Summary = x.Values.Select(y => new { Total = y.Value }).AsEnumerable()
                })
                .ToListAsync();
            Console.WriteLine($"items3: {JsonSerializer.Serialize(items3)}");
```

will throw an exception:
```
An unhandled exception of type 'System.InvalidOperationException' occurred in System.Private.CoreLib.dll
Operation is not valid due to the current state of the object
```
This is just a small example of what I'm actually  trying to do, I'm trying to select an association but also using it in orderBy, separately they work fine, but together they do not.
I encountered this issue working on linq2db v3.3.0, PostgreSQL database and .NET 5.0.
Before reporting this issue I created a this example above and using Microsoft SQL Server, tried a few versions of linq2db from 3.0 up to latest 3.4.4, also tried switching to .NET Core 3.1 and the same thing happens. 
I have attached the full example with database schema for SQL Server.

### Steps to reproduce
Open the provided solution in attachment and run it.
First two queries will run fine but the third one will throw an exception.

### Environment details
linq2db version: 3.4.4
Database Server: Microsoft SQL Server Express 13.0.5103
Database Provider:  SQL Server
Operating system: Windows 10
.NET Framework: 5.0

[LinqToDBExample.zip](https://github.com/linq2db/linq2db/files/7223281/LinqToDBExample.zip)

**Comment by darko1979** [comment]
So this is a bit awkward, I was trying to make this work on my project for at least an hour, and today testing and preparing an example for an hour, and as soon as I submitted the issue I had an idea to test and apparently this works:
```c#
            var items3 = await db.Items
                .OrderBy(x => x.Values.Sum(y => (decimal?)y.Value) ?? (decimal)0.0)
                .Select(x => new {
                    Id = x.Id,
                    Text = x.Text,
                    Summary = x.Values.Select(y => new { Total = y.Value }).AsEnumerable()
                })
                .ToListAsync();
```
Not sure anymore is this still an issue to be resolved.

**Comment by sdanyliv** [comment]
It is real query or just extraction of problem? Such calculations better to do on the client side, because you load all data. Eager Loaging here just slowdowns query.

**Comment by darko1979** [comment]
It's an approximation of what I'm doing. I have users and transactions summary table that is already summarized transaction table by transaction type, I have to return the summarized values but also sort users by these summarized values.

**Comment by MaceWindu** [comment]
Was fixed in 5.1.0
